[Dangabit] iP

Dangabit opened the pull request:
# KiraBot

_Manage your tasks in a better way!_

**KiraBot lets you record down tasks and have several features to make your life so much better!**

### Why KiraBot
- text-based
- easy to use
- ~~FAST~~ LIGHTNING MCQUEEN
- FREE. No hidden charges!

### HOW DO I GET IT

_5 Simple Steps_
1. Find it from [here](https://github.com/Dangabit/ip/releases).
2. Download the lastest jar release.
3. Double-click KiraGUI.jar or run `java -jar KiraCLI.jar` 
4. Add your tasks.
5. Let it manage your tasks for you 😉

### Features

- [x] Manages Tasks as ToDo, Deadlines, Events!
- [x] Tells you what you have on TODAY!
- [x] Now with a GUI!
- [ ] Reminders (coming soon)

### I love coding
Sure you do. If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class KiraBot { // For CLI
    public static void main(String[] args) {
        KiraBot bot = new KiraBot();
        bot.run();
    }
}
```
```
public class KiraLauncher { // For GUI
    public static void main(String[] args) {
        Application.launch(KiraApp.class, args);
    }
}
```

> Best task management CLI ever ~NotDangabit
> Best task management GUI ever ~AlsoNotDangabit

Commit:
Add Gradle support

Dangabit committed:
Implemented Greet, Echo, Quit

Dangabit committed:
Implemented Add, List

Dangabit committed:
Implements Mark as Done

Dangabit committed:
Implement ToDo, Event, Deadline

Dangabit committed:
Implement Automated Text UI Testing

Dangabit committed:
Implement Error Handling

There are several problems when the user input some invalid data.

KiraException will handle all the exception caused by this software.
Currently, the exceptions checked are invalid-index when marking or
unmarking, invalid command and storage out-of-bounds.

Error when parsing user input will be dealt with when Parser is
implemented.

Dangabit committed:
Implement Delete

KiraBot can now delete tasks from the list. Furthermore, the list is now
expandable. The limit exception has been removed.

Dangabit committed:
Use Enums

Dangabit committed:
Implement save function

The bot requires a saving function to save the session of the user.

Saving is done at every command. The save file is in csv for easy
parsing and loading.

Dangabit committed:
Implement loading function

When the bot starts up, it should read the save file, if any, and load
the state of the database.

KiraBot now reads the file and load in the saved state of the task list.
Further changes are made to change the format of the delimiters to allow
for commas in task description.

Dangabit committed:
Change to DateTime for tasks

Currently, all relevant tasks with a date is stored as String. Now, a
specific format is required to save the dates and time for the tasks.

Additionally, a new command TODAY is implemented to retrieve all tasks
that are within the time frame of today.

Dangabit committed:
Merge branch-level-7 to master

Dangabit committed:
Refactor code to include more OOP and packages

Currently, there is no organisation for the different java files and no
dedicated class for each logic abstraction.

More classes have been implemented to separate the logic and taskList no
longer return formatted strings for each method as it is not the job.
Furthermore, Parser will manage all of the error during parsing of the
commands.

Dangabit committed:
Merge branch 'add-gradle-support' of https://github.com/Dangabit/ip into add-gradle-support

Dangabit committed:
Merge branch 'add-gradle-support'

Dangabit committed:
Automate with Gradle, add JUnit tests

Dangabit committed:
Fix corrupted file load

Dangabit committed:
Remove redundent codes

Dangabit committed:
Fix test case

Dangabit committed:
Refactor commands and write JavaDocs

Dangabit committed:
Tweak code to comply with standard

Dangabit committed:
Implement find function

With a lot of tasks, users might have a hard time finding some specific
task.

Therefore, a find command has been implemented for the user to use a
particular keyword to find all the tasks matching it.

Dangabit committed:
Resolve merge conflicts

Dangabit committed:
Merge branch 'master' of https://github.com/Dangabit/ip

Dangabit committed:
Update README.md

Dangabit committed:
Fix style according to Checkstyle

Dangabit committed:
Merge branch 'master' of https://github.com/Dangabit/ip

Dangabit committed:
Setup GUI

lyndonlim27 commented on src/main/java/kira/KiraBot.java:
```diff
@@ -0,0 +1,65 @@
+
+    private void listenForCommand() {
+        Scanner scanner = new Scanner(System.in);
+        boolean isActive = true;
```
Great! You named the boolean variable such that it sounds like a question

lyndonlim27 commented on src/main/java/kira/command/FindCommand.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public boolean execute(Ui ui, TaskList tasklist) {
+        ui.findMsg(tasklist.findByKey(this.keyword), this.keyword);
```
From tutorial 1,  I learnt that if your method does not have a parameter that has an identical name to your class variable (e.g keyword). It is not required to include "this". However, it is optional. Just wanted to let you know.

lyndonlim27 commented on src/main/java/kira/storage/SaveLoad.java:
```diff
@@ -0,0 +1,105 @@
+     * @throws KiraException File-Not-Found
+     */
+    public static List<Task> load(String pathname) throws KiraException {
+        List<Task> taskList = new ArrayList<>();
```
```suggestion
        List<Task> taskLists = new ArrayList<>();
```
As this is a list which represents a collection of objects, plural form should be used. I noticed the same instances in several other places too.

lyndonlim27 commented on src/main/java/kira/task/Deadline.java:
```diff
@@ -0,0 +1,64 @@
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+            this.deadline = LocalDateTime.parse(deadline, formatter);
+        } catch (DateTimeParseException e) {
+            throw new KiraException(
+                    "Please input your date by this format:"
+                    + " yyyy-MM-dd HHmm");
```
Great! You used the correct indentation for wrapped lines

lyndonlim27 reviewed (commented):
Well done overall! Most coding standards have been met except for a few i pointed out. Your javadocs is great too!

Dangabit committed:
Improve GUI functionality

Dangabit committed:
Fix style error

bipbipboopboop commented on src/main/java/kira/KiraBot.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    private void listenForCommand() {
+        Scanner scanner = new Scanner(System.in);
```
You can use try-with-resource here, Java will automatically close the Scanner for you.

bipbipboopboop commented on src/main/java/kira/ui/Parser.java:
```diff
@@ -0,0 +1,148 @@
+     * @throws KiraException Incorrect-Format, Invalid-Index
+     */
+    public static Command parse(String input) throws KiraException {
+        String[] temp = input.split(" ", 2);
```
You should rename this variable to a more meaningful name.

bipbipboopboop commented on src/main/java/kira/ui/Ui.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    public String getMessage() {
+        return this.message;
```
`this` is not necessary here. In this module, we are recommended to not use that if possible.

bipbipboopboop commented on src/main/java/kira/command/ByeCommand.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.concurrent.CompletableFuture;
+
+import javafx.application.Platform;
+import kira.storage.TaskList;
+import kira.ui.Ui;
```
May be split this import by an empt line?

bipbipboopboop reviewed (commented):
Your code looks good, it is quite pleasant to read.

Dangabit committed:
Fix several issues

As part of the cleanup and ensuring KiraBot GUI jar can be compiled,
several changes have been made.
1. README.md has been updated to include GUI functionality
2. build.gradle is modified to enable assertions for future update
3. KiraBot GUI can now display the load file warning and greet
4. Slight modifications to other files to align with coding standards

Dangabit committed:
Implement dual build for CLI and GUI

Dangabit committed:
Enable and implement assertions

Dangabit committed:
Refactor code

Some codes are currently of poor quality, hence are refactored.
Changes are made to better follow SLAP and reduce arrowhead style.

Dangabit committed:
Merge pull request #1 from Dangabit/branch-A-CodeQuality

Improve Code Quality

Dangabit committed:
Add assert on parser

Dangabit committed:
Merge pull request #2 from Dangabit/branch-A-Assertions

Enable and implement assertions

Dangabit committed:
Implement help command

Newer users might not know how to use some command. Therefore, there is
a need for a help command that informs the user on the proper format and
usage of the commands.

Dangabit committed:
Update docs

Dangabit committed:
Fix README formatting
